### PR TITLE
mktime64 test fixes, and optimizations

### PR DESCRIPTION
**Pull request by rurban**
merge and fix rdiez' improvements.
improve the Makefile, detect Linux and Darwin, and add the needed HAS_TM defines.
  support make DEBUG=1 and USE_TM64=1
add tap.h declarations
replace %lld printf with %PRId64 from inttypes.h
Fix fragile t/mktime64.t
  use an hour which does not overflow to the next or prev day.
  test all kind of years, esp. leap years with out of bound dates

various minor optimizations

### CODE REVIEW

**Comment by jakesyl** [comment]
+1 on merge.

File: t/mktime64.t.c [hunk 1]
```diff
@@ -50,6 +49,12 @@ int main(void) {
     date.tm_mday = 37;
     date.tm_wday = 9;           /* deliberately wrong week day */
     date.tm_yday = 487;         /* and wrong year day */
+#ifdef HAS_TM_TM_GMTOFF
+    /*date.tm_gmtoff = 0;*/
```
Comment (schwern) [inline t/mktime64.t.c]: Should this be uncommented?

File: t/tap.h [hunk 2]
```diff
@@ -1,6 +1,12 @@
 #include <stdarg.h>
 #include "time64.h"
 
+#ifdef USE_TM64
```
Comment (schwern) [inline t/tap.h]: Maybe we need to define a year_t in the library itself.

File: t/tap.c [hunk 3]
```diff
@@ -87,8 +86,7 @@ int is_str(const char* have, const char* want, const char *message, ...) {
     do_test( test, message, args );
 
     if( !test ) {
-        diag("have: %s", have);
-        diag("want: %s", want);
+      diag("have: %s, want: %s", have, want);
```
Comment (schwern) [inline t/tap.c]: have and want are supposed to be on two lines, the two values are lined up to more easily see the difference.

File: t/tap.c [hunk 4]
```diff
@@ -106,8 +104,7 @@ int is_Int64(const Int64 have, const Int64 want, const char *name, ...) {
     do_test( test, name, args );
 
     if( !test ) {
-        diag("have: %"PRId64, have);
-        diag("want: %"PRId64, want);
+        diag("have: %"PRId64" want: %"PRId64" diff: %"PRId64, have, want, have-want);
```
Comment (schwern) [inline t/tap.c]: Ditto.

File: Makefile [hunk 5]
```diff
@@ -20,7 +20,8 @@ ifeq ($(UNAME_S),Linux)
   # (instead of __tm_gmtoff and __tm_zone) in struct tm,
   # and _POSIX_SOURCE (there are alternatives) for tzset().
   # _BSD_SOURCE is deprecated, use _DEFAULT_SOURCE instead
-  CFLAGS += -D_DEFAULT_SOURCE -D_POSIX_SOURCE -DHAS_TM_TM_GMTOFF -DHAS_TM_TM_ZONE
+  CFLAGS += -D_BSD_SOURCE -D_DEFAULT_SOURCE -D_POSIX_SOURCE
+  #CFLAGS += -D_DEFAULT_SOURCE -D_POSIX_SOURCE -DHAS_TM_TM_GMTOFF -DHAS_TM_TM_ZONE
```
Comment (schwern) [inline Makefile]: Reminding myself to remove this commented out line.

**Comment by schwern** [comment]
Thanks a lot for doing all that work! Sorry it took so long to review and merge.

**Comment by rurban** [comment]
> What's a .i file for?

It's the naming convention for cc -E preprocessed .c files. You inspect macro expansions and included headers.

**Comment by rurban** [comment]
Welcome back! We missed you

**Pull request merged**